### PR TITLE
[feat]:[pl-26313]: Integrate DELETE Template Flow with GitX

### PR DESCRIPTION
**Pull request by adivishy1**
## Describe your changes
Integrate DELETE Template Flow with GitX. The following flows have been tested

API deleteTemplateVersionsOfIdentifier
New GitX 
**Remote**
single template version - done
multiple versions - done
**Inline**
single template version - done
multiple versions - done
**Old GitSync**
single template version - done
multiple versions - done

API deleteTemplateVersion									
New GitX 
**Remote**
single template version - done
**Inline**
single template version - done
**Old GitSync**
single template version - done

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

---
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/35942)

1.

### CODE REVIEW

**Comment by mohitg0795** [comment]
*[`840-template-service/src/main/java/io/harness/repositories/NGTemplateRepositoryCustomImpl.java` line 342 at r2](https://reviewable.io/reviews/harness/harness-core/35942#-N8TTNYwDsKsm1Wu0e3v:-N8TTNYwDsKsm1Wu0e3w:bj3ayui) ([raw file](https://github.com/harness/harness-core/blob/0c29625ab78cb33309a419a5c05db440c9710392/840-template-service/src/main/java/io/harness/repositories/NGTemplateRepositoryCustomImpl.java#L342)):*
> ```Java
>   }
> 
>   private Criteria buildCriteriaForDelete(TemplateEntity templateEntity, boolean deleted) {
> ```

whats the logic for adding "deleted" as argument?

**Comment by mohitg0795** [comment]
*[`840-template-service/src/main/java/io/harness/repositories/NGTemplateRepositoryCustomImpl.java` line 345 at r2](https://reviewable.io/reviews/harness/harness-core/35942#-N8TTTat5s8nOfiksLI2:-N8TTTauAiHgQAEb-6My:b6r179p) ([raw file](https://github.com/harness/harness-core/blob/0c29625ab78cb33309a419a5c05db440c9710392/840-template-service/src/main/java/io/harness/repositories/NGTemplateRepositoryCustomImpl.java#L345)):*
> ```Java
>     return Criteria.where(TemplateEntityKeys.accountId)
>         .is(templateEntity.getAccountId())
>         .where(TemplateEntityKeys.deleted)
> ```

why is it just randomly added in the middle? I mean fields should be in some order right? eg: accountid, orgid, then project id and so on.

**Comment by mohitg0795** [comment]
*[`840-template-service/src/main/java/io/harness/template/services/NGTemplateServiceImpl.java` line 413 at r2](https://reviewable.io/reviews/harness/harness-core/35942#-N8TTnGt7iH0yGc_4ad3:-N8TTnGt7iH0yGc_4ad4:b-scvkfq) ([raw file](https://github.com/harness/harness-core/blob/0c29625ab78cb33309a419a5c05db440c9710392/840-template-service/src/main/java/io/harness/template/services/NGTemplateServiceImpl.java#L413)):*
> ```Java
>         }
>       } else {
>         templateRepository.deleteTemplate(templateToDelete, comments);
> ```

we should have similar usecase of "isHardDeleteEntitiesFeatureFlagEnabled" here as well right?

**Review by adivishy1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/35942)* status: 4 file(s) to review. 3. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`840-template-service/src/main/java/io/harness/repositories/NGTemplateRepositoryCustomImpl.java` line 342 at r2](https://reviewable.io/reviews/harness/harness-core/35942#-N8TTNYwDsKsm1Wu0e3v:-N8TTcXjB3s_04BSWmKC:bbnmuqw) ([raw file](https://github.com/harness/harness-core/blob/0c29625ab78cb33309a419a5c05db440c9710392/840-template-service/src/main/java/io/harness/repositories/NGTemplateRepositoryCustomImpl.java#L342)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

whats the logic for adding "deleted" as argument?
</blockquote></details>

basically, the template entity shouldnt be deleted prior to delete operation being performed
___
*[`840-template-service/src/main/java/io/harness/repositories/NGTemplateRepositoryCustomImpl.java` line 345 at r2](https://reviewable.io/reviews/harness/harness-core/35942#-N8TTTat5s8nOfiksLI2:-N8TU7w5Ckv2OgkCSDxX:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/0c29625ab78cb33309a419a5c05db440c9710392/840-template-service/src/main/java/io/harness/repositories/NGTemplateRepositoryCustomImpl.java#L345)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

why is it just randomly added in the middle? I mean fields should be in some order right? eg: accountid, orgid, then project id and so on.
</blockquote></details>

Done.
___
*[`840-template-service/src/main/java/io/harness/template/services/NGTemplateServiceImpl.java` line 413 at r2](https://reviewable.io/reviews/harness/harness-core/35942#-N8TTnGt7iH0yGc_4ad3:-N8TUAod7P4Q8Gs2zbzR:bbyuzwc) ([raw file](https://github.com/harness/harness-core/blob/0c29625ab78cb33309a419a5c05db440c9710392/840-template-service/src/main/java/io/harness/template/services/NGTemplateServiceImpl.java#L413)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

we should have similar usecase of "isHardDeleteEntitiesFeatureFlagEnabled" here as well right?
</blockquote></details>

For hard delete, we need to delete the template entity from git, in the new gitX flow, we dont that right, also pipeline team is also following the same approach for delete and not deleting from git for new gitX

**Comment by mohitg0795** [comment]
*[`840-template-service/src/main/java/io/harness/repositories/NGTemplateRepositoryCustomImpl.java` line 342 at r2](https://reviewable.io/reviews/harness/harness-core/35942#-N8TTNYwDsKsm1Wu0e3v:-N8VKbMAD9Ia1nG4Fac4:bfjaulk) ([raw file](https://github.com/harness/harness-core/blob/0c29625ab78cb33309a419a5c05db440c9710392/840-template-service/src/main/java/io/harness/repositories/NGTemplateRepositoryCustomImpl.java#L342)):*
<details><summary><i>Previously, adivishy1 (Adithya Viswanathan) wrote…</i></summary><blockquote>

basically, the template entity shouldnt be deleted prior to delete operation being performed
</blockquote></details>

I mean if this is criteria for deleting, then why is this argument required?

**Review by adivishy1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/35942)* status: 4 file(s) to review. 1. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`840-template-service/src/main/java/io/harness/repositories/NGTemplateRepositoryCustomImpl.java` line 342 at r2](https://reviewable.io/reviews/harness/harness-core/35942#-N8TTNYwDsKsm1Wu0e3v:-N8Wwchf8VXQMrdHdFFp:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/0c29625ab78cb33309a419a5c05db440c9710392/840-template-service/src/main/java/io/harness/repositories/NGTemplateRepositoryCustomImpl.java#L342)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

I mean if this is criteria for deleting, then why is this argument required?
</blockquote></details>

Done.

**Review by inderschera — CHANGES_REQUESTED** [review]
> Reviewed 2 of 3 files at r1, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/35942)* status: 1 change requests. 1. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`840-template-service/src/main/java/io/harness/template/services/NGTemplateServiceImpl.java` line 407 at r4](https://reviewable.io/reviews/harness/harness-core/35942#-N8XQipL-HqzYAPdbCkx:-N8XQipO7Ffar_TTSCn4:b-u4hovl) ([raw file](https://github.com/harness/harness-core/blob/c53adc02e52ac6c52bf095f454e9951d241f99c4/840-template-service/src/main/java/io/harness/template/services/NGTemplateServiceImpl.java#L407)):*
> ```Java
>             return true;
>           } else {
>             throw new InvalidRequestException(format(
> ```

Let's move it to NgTemplateServiceHelper



_Code quote:_
```Java
      if (templateServiceHelper.isOldGitSync(templateToDelete)) {
        if (isHardDeleteEntitiesFeatureFlagEnabled(accountId)) {
          templateRepository.hardDeleteTemplateForOldGitSync(templateToDelete, comments);
          return true;
        } else {
          TemplateEntity deletedTemplate = templateRepository.deleteTemplateForOldGitSync(withDeleted, comments);
          if (deletedTemplate.getDeleted()) {
            return true;
          } else {
            throw new InvalidRequestException(format(
                "Template with identifier [%s] and versionLabel [%s], under Project[%s], Organization [%s] couldn't be deleted.",
                templateIdentifier, versionLabel, projectIdentifier, orgIdentifier));
          }
        }
      } else {
        templateRepository.deleteTemplate(templateToDelete, comments);
        return true;
      }
```
___
*[`840-template-service/src/main/java/io/harness/template/services/NGTemplateServiceImpl.java` line 581 at r4](https://reviewable.io/reviews/harness/harness-core/35942#-N8XRMctBAQuPInKEAXb:-N8XRMctBAQuPInKEAXc:b-3ne0ja) ([raw file](https://github.com/harness/harness-core/blob/c53adc02e52ac6c52bf095f454e9951d241f99c4/840-template-service/src/main/java/io/harness/template/services/NGTemplateServiceImpl.java#L581)):*
> ```Java
>           } else {
>             templateRepository.deleteTemplate(templateEntity, "");
>           }
> ```

We don't need this check here. It's already being checked earlier.



_Code quote:_
```Java
          if (templateServiceHelper.isOldGitSync(templateEntity)) {
            templateRepository.hardDeleteTemplateForOldGitSync(templateEntity, "");
          } else {
            templateRepository.deleteTemplate(templateEntity, "");
          }
```

**Comment by diptiman-adak** [comment]
Quality gate Quality gate: **_Passed_** 
 


            
 #### 1 Issues 
* Anti-pattern 1 Anti-patterns 

 [Details](https://app.embold.io/organization/gh/harness/repositories/7752dbb3099d2dba3b5d37f5931a5e4c/pull-request/35942)

**Review by mohitg0795 — APPROVED** [review]
> Reviewed 1 of 3 files at r1, 2 of 2 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/35942)* status: :shipit: complete! ✓

**Review by adivishy1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/35942)* status: 5 file(s) to review. 1 change requests. 2. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`840-template-service/src/main/java/io/harness/template/services/NGTemplateServiceImpl.java` line 407 at r4](https://reviewable.io/reviews/harness/harness-core/35942#-N8XQipL-HqzYAPdbCkx:-N8YdVyUBxkY1yG8qYCD:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/c53adc02e52ac6c52bf095f454e9951d241f99c4/840-template-service/src/main/java/io/harness/template/services/NGTemplateServiceImpl.java#L407)):*
<details><summary><i>Previously, inderschera (Inderpreet Chera) wrote…</i></summary><blockquote>

Let's move it to NgTemplateServiceHelper
</blockquote></details>

Done.
___
*[`840-template-service/src/main/java/io/harness/template/services/NGTemplateServiceImpl.java` line 581 at r4](https://reviewable.io/reviews/harness/harness-core/35942#-N8XRMctBAQuPInKEAXb:-N8YTQEE4umU9ZrR_Wj4:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/c53adc02e52ac6c52bf095f454e9951d241f99c4/840-template-service/src/main/java/io/harness/template/services/NGTemplateServiceImpl.java#L581)):*
<details><summary><i>Previously, inderschera (Inderpreet Chera) wrote…</i></summary><blockquote>

We don't need this check here. It's already being checked earlier.
</blockquote></details>

Done.

**Review by bhavya181 — DISMISSED** [review]
> Reviewed 1 of 3 files at r1, 2 of 2 files at r4, 1 of 4 files at r6, 4 of 4 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/35942)* status: 1 change requests. 2. unresolved discussions.

**Review by mohitg0795 — APPROVED** [review]
:lgtm:

> Reviewed 1 of 4 files at r6, 4 of 4 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/35942#-:-N8Yh64e-u72jRWm6hv7:bnfp4nl)* status: 1 change requests. 2. unresolved discussions.

**Review by bhavya181 — APPROVED** [review]
> Reviewed 6 of 6 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/35942)* status: 1 change requests. 2. unresolved discussions.

**Review by inderschera — APPROVED** [review]
> Reviewed 2 of 4 files at r7, 6 of 6 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/35942)* status: :shipit: complete! ✓
